src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/RemoveUnusedLocals.Fixer.cs

```diff
@@ -1,27 +1,174 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Composition;
+using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis.FindSymbols;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CodeActions;
 
 namespace Microsoft.CodeQuality.Analyzers.Maintainability
 {
     /// <summary>
     /// CA1804: Remove unused locals
     /// </summary>
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = AvoidUnusedPrivateFieldsAnalyzer.RuleId), Shared]
     public abstract class RemoveUnusedLocalsFixer : CodeFixProvider
     {
-        public sealed override FixAllProvider GetFixAllProvider()
+        internal const string RuleId = "CA1804";
+        private readonly NodesProvider _nodesProvider;
+
+        protected RemoveUnusedLocalsFixer(NodesProvider nodesProvider) => _nodesProvider = nodesProvider;
+
+        public sealed override FixAllProvider GetFixAllProvider() => new RemoveLocalFixAllProvider(_nodesProvider);
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            Diagnostic diagnostic = context.Diagnostics.Single();
+            var nodesToRemove = await _nodesProvider.GetNodesToRemoveAsync(context.Document, diagnostic, context.CancellationToken).ConfigureAwait(false);
```

Comment:
Is there a reason why we compute this outside of the code action? We generally execute some code outside of the code action if the result of that computation will lead us to bail out early without even registering a code fix. Otherwise, it is recommend to move all this code into the code action itself.

Reply:
Note the reason for this is perf. Code executed within a code action is only executed when user attempts to apply a fix, which the code is executed every time we attempt to compute available code fixes for a given diagnostic (such as pressing Ctrl + Dot)

Reply:
Thank you! Makes sense. The reason is in other fixes such as AvoidUnusedPrivateFields. They follow the same pattern. I will try to move perf related method within code action. Maybe we should review other fixers too.

---
In reply to: [152388172](https://github.com/dotnet/roslyn-analyzers/pull/1336#discussion_r152388172) [](ancestors = 152388172)